### PR TITLE
Bluetooth: Initial conversion from bt_storage to the settings API

### PR DESCRIPTION
**Pull request by jhedberg**
This is an initial patch set to transition from the bt_storage API to a settings-based solution. It contains support for local values such as identity address and IRK, as well as remote pairing-related keys (the bt_keys struct). In follow-up PRs I'll work on adding support for per-device CCC as well as support for mesh storage.

### CODE REVIEW

File: subsys/bluetooth/host/settings.c
Comment (jhedberg) [inline subsys/bluetooth/host/settings.c]: Note about this: I'm not 100% sure if the settings subsystem ensures that this API can be called multiple times (potentially by multiple subsystems). At least parts of it were protected by an "is_inited" boolean, but not all, I think. This might be good to fix in settings. Alternatively, we could move the requirement of calling settings_subsys_init() to the application side, i.e. before calling bt_enable(). That would make it consistent with the requirement of calling settings_load() on the application side as well. Thoughts?

**Review by carlescufi — DISMISSED** [review]

File: subsys/bluetooth/host/settings.c
Comment (jhedberg) [reply subsys/bluetooth/host/settings.c]: Before doing that however, we should fix settings so that settings_register() either cleanly fails or cleanly succeeds if called before settings_subsys_init() has been called. Currently that's not that case since settings_register() will always succeed, however settings_subsys_init() will clear the list of handlers if it's called later. I'd propose using static initialization of the handlers list, so that the sys_slist_init() call in settings_store_init() is not needed. That way it doesn't matter in which order things happen, and we can always safely call settings_register().

That said, I think I might still prefer for bt_enable() to fail if the application has not called settings_subsys_init(), which means that settings_register() should return failure in that case. Any thoughts @nvlsianpu ?

Comment (jhedberg) [reply subsys/bluetooth/host/settings.c]: One more thought: if we move the settings handlers to a dedicated ROM linker section, the settings_register() API will go away. That would change things a bit. One of the reasons I did settings_load() outside of bt_enable() is that with mesh the mesh settings handlers would only be registered upon bt_mesh_init() (which is called after bt_enable). However, if we have a dedicated linker sections, all handlers are already there from the very beginning, making it safer to call settings_load() at any time (it might still not be completely safe though since there might still be run-time init functions that some subsystems need to get called first)

Comment (jhedberg) [reply subsys/bluetooth/host/settings.c]: For the record: I've just pushed a third patch to this PR, showing how I'd do a ROM linker section for the Bluetooth-internal submodule handlers. The solution for the settings handlers would look quite similar.

File: subsys/bluetooth/host/hci_core.c
Comment (jhedberg) [reply subsys/bluetooth/host/hci_core.c]: Where exactly? Note that bt_unpair() is essentially the old bt_storage_clear() with the bt_storage->clear() called removed and moved to a new position in hci_core.c that's inside a `#if defined(CONFIG_BT_CONN)` section. Currently the API does not do any persistent clearing since there's no support for persistent pairing storage. As soon as it's added the API will do it, i.e. it'll then clear both runtime and persistent info, whereas currently it only clears runtime info (which is the only info we have for now).

File: subsys/bluetooth/host/settings.c
Comment (nvlsianpu) [reply subsys/bluetooth/host/settings.c]: Agree, but have in mind that fcb test `tests/subsys/settings/fcb/src/settings_test_fcb.c'` need to be aligned as settings_register() is called without initaialization proir (any test_config_insert* testcase)

File: subsys/bluetooth/host/hci_core.c
Comment (carlescufi) [reply subsys/bluetooth/host/hci_core.c]: Of course, I failed to see that here. Disregard.

File: subsys/bluetooth/host/settings.c
Comment (carlescufi) [reply subsys/bluetooth/host/settings.c]: Agree with the dependency on fixing `settings_register` before we can actually safely consider the initialization path. So should we put this PR on hold while we send a PR to fix the settings subsystem itself?

Comment (nvlsianpu) [reply subsys/bluetooth/host/settings.c]: Think not necessarily hood on this - as can be implemented properly even now.

**Comment by jhedberg** [comment]
I've now also pushed patches to add support for pairing keys storage. I also discovered that I was badly overflowing the Bluetooth RX stack as well as the system workqueue stack due to the settings API calls. So both of those stacks are now 2k by default (I've updated my first patch to reflect this).

**Review by dbkinder — APPROVED** [review]
Doc change LGTM, thanks.

**Comment by jhedberg** [comment]
Any other comments on this PR? The "who should select MPU_ALLOW_FLASH_WRITE" is still an open issue (perhaps it should be SETTINGS rather than BT_SETTINGS), but I think that can be fixed up in later pull requests as well. It'd be good to get this merged since I'm starting to have other related patches queued up locally already (e.g. for mesh). From github policy perspective the review from @dbkinder is sufficient for merging, but it's only about the documentation, i.e. it'd be good to get a +1 from a functional perspective as well.

**Pull request merged**
